universe/src/server.rs

```diff
@@ -0,0 +1,210 @@
+use std::{
+    net::SocketAddr,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use common::{
+    host_info::{self, HostInfo},
+    region::{self, Region, Zone},
+};
+use pin_project::{pin_project, pinned_drop};
+use proto::universe::universe_client::UniverseClient;
+use proto::universe::universe_server::Universe;
+use proto::universe::{
+    CreateKeyspaceRequest, CreateKeyspaceResponse, ListKeyspacesRequest, ListKeyspacesResponse,
+};
+use tokio::sync::broadcast;
+use tokio_stream::{
+    wrappers::{errors::BroadcastStreamRecvError, BroadcastStream},
+    Stream,
+};
+use tokio_util::sync::CancellationToken;
+use tonic::{Request, Response, Status};
+use tracing::{debug, info, instrument};
+
+use crate::storage::{
+    cassandra::{self, Cassandra},
+    Storage,
+};
+
+/// Implementation of the Universe manager.
+pub struct UniverseServer<S: Storage> {
+    storage: Arc<S>,
+}
+
+impl<S: Storage> UniverseServer<S> {
+    /// Creates a new UniverseServer.
+    pub fn new(storage: Arc<S>) -> Self {
+        Self { storage }
+    }
+}
+
+#[tonic::async_trait]
+impl<S: Storage> Universe for UniverseServer<S> {
+    #[instrument(skip(self))]
+    async fn create_keyspace(
+        &self,
+        request: Request<CreateKeyspaceRequest>,
+    ) -> Result<Response<CreateKeyspaceResponse>, Status> {
+        debug!("Got a create_keyspace request: {:?}", request);
+
+        let req_inner = request.into_inner();
+        let keyspace_id = self
+            .storage
+            .create_keyspace(
+                &req_inner.name,
+                &req_inner.namespace,
+                req_inner.primary_zone,
+                req_inner.base_key_ranges,
+            )
+            .await
+            .map_err(|e| Status::internal(format!("Failed to create keyspace: {}", e)))?;
+
+        let response = CreateKeyspaceResponse { keyspace_id };
+        Ok(Response::new(response))
+    }
+
+    #[instrument(skip(self))]
+    async fn list_keyspaces(
+        &self,
+        request: Request<ListKeyspacesRequest>,
```

Comment:
This takes a region, right? we should filter the results by primary zone. OK if that's in memory right now